pkg/migration/migrations/public_schema_migration.go

```diff
@@ -0,0 +1,203 @@
+			ParentID:   dbID,
+			Name:       "public",
+			ID:         publicSchemaID,
+			Privileges: publicSchemaPrivileges,
+			Version:    1,
+		}).BuildCreatedMutableSchema()
+
+		descID := publicSchemaDesc.GetID()
+		idKey := catalogkeys.MakeSchemaNameKey(d.Codec, dbID, publicSchemaDesc.GetName())
+
+		key := catalogkeys.EncodeNameKey(d.Codec, catalogkeys.NewNameKeyComponents(dbID, keys.RootNamespaceID, tree.PublicSchema))
+
+		b := txn.NewBatch()
+
+		// Remove namespace entry for old public schema.
+		b.Del(key)
+		b.CPut(idKey, descID, nil)
+		if err := catalogkv.WriteNewDescToBatch(
+			ctx,
+			false,
+			d.Settings,
+			b,
+			d.Codec,
+			descID,
+			publicSchemaDesc,
+		); err != nil {
+			return err
+		}
+
+		dbDesc.DescriptorProto().GetDatabase().Schemas = map[string]descpb.DatabaseDescriptor_SchemaInfo{
+			tree.PublicSchema: {
+				ID: publicSchemaID,
+			},
+		}
+		dbDesc.MaybeIncrementVersion()
+		if err := catalogkv.WriteDescToBatch(
+			ctx, false, d.Settings, b, d.Codec, dbDesc.GetID(), dbDesc,
+		); err != nil {
+			return err
+		}
+
+		if err := migrateObjectsInDatabase(ctx, dbID, d, txn, b, publicSchemaID, descriptors); err != nil {
+			return err
+		}
+
+		return txn.Run(ctx, b)
+	})
+}
+
+func migrateObjectsInDatabase(
+	ctx context.Context,
+	dbID descpb.ID,
+	d migration.TenantDeps,
+	txn *kv.Txn,
+	batch *kv.Batch,
+	newPublicSchemaID descpb.ID,
+	descriptors *descs.Collection,
+) error {
+	descs, err := descriptors.GetAllDescriptorsInDatabase(ctx, txn, dbID)
+	if err != nil {
+		return err
+	}
+
+	var modifiedDescs []catalog.MutableDescriptor
+	for _, desc := range descs {
+		b := catalogkv.NewBuilder(desc.DescriptorProto())
+		switch b.DescriptorType() {
+		case catalog.Database, catalog.Schema:
+			panic(fmt.Sprintf("unexpected descriptor type %v", b.DescriptorType()))
+		case catalog.Table:
+			table := b.BuildExistingMutable()
+			oldKey := catalogkeys.MakeObjectNameKey(d.Codec, table.GetParentID(), table.GetParentSchemaID(), table.GetName())
+			batch.Del(oldKey)
+			table.DescriptorProto().GetTable().UnexposedParentSchemaID = newPublicSchemaID
+			newKey := catalogkeys.MakeObjectNameKey(d.Codec, table.GetParentID(), table.GetParentSchemaID(), table.GetName())
+			batch.Put(newKey, table.GetID())
+			modifiedDescs = append(modifiedDescs, table)
+		case catalog.Type:
+			typ := b.BuildExistingMutable()
+			oldKey := catalogkeys.MakeObjectNameKey(d.Codec, typ.GetParentID(), typ.GetParentSchemaID(), typ.GetName())
+			batch.Del(oldKey)
+			typ.DescriptorProto().GetType().ParentSchemaID = newPublicSchemaID
+			newKey := catalogkeys.MakeObjectNameKey(d.Codec, typ.GetParentID(), typ.GetParentSchemaID(), typ.GetName())
+			batch.Put(newKey, typ.GetID())
+			modifiedDescs = append(modifiedDescs, typ)
+		}
+	}
+	for _, modified := range modifiedDescs {
+		err := catalogkv.WriteDescToBatch(
+			ctx, false, d.Settings, batch, d.Codec, modified.GetID(), modified,
+		)
```

Comment:
Here also, use `descriptors.WriteDescToBatch` instead if possible, it should be better and safer?

Reply:
Done